amr-wind/physics/TerrainDrag.cpp

```diff
@@ -0,0 +1,172 @@
+                    const amrex::Real x1 = prob_lo[0] + (i + 0.5) * dx[0];
+                    const amrex::Real x2 = prob_lo[1] + (j + 0.5) * dx[1];
+                    const amrex::Real x3 = prob_lo[2] + (k + 0.5) * dx[2];
+                    // Terrain Height
+                    amrex::Real residual = 10000;
+                    amrex::Real terrainHt = 0.0;
+                    for (unsigned ii = 0; ii < terrainSize; ++ii) {
+                        const amrex::Real radius = std::sqrt(
+                            std::pow(x1 - xterrain_ptr[ii], 2) +
+                            std::pow(x2 - yterrain_ptr[ii], 2));
+                        if (radius < residual) {
+                            residual = radius;
+                            terrainHt = zterrain_ptr[ii];
+                        }
+                        if (radius < dx[0]) {
+                            break;
+                        }
+                    }
+                    const amrex::Real turnOn = (x3 <= terrainHt) ? 1.0 : 0.0;
+                    levelBlanking(i, j, k, 0) = turnOn;
+                    residual = 10000;
+                    amrex::Real roughz0 = 0.1;
+                    for (unsigned ii = 0; ii < roughnessSize; ++ii) {
+                        const amrex::Real radius = std::sqrt(
+                            std::pow(x1 - xrough_ptr[ii], 2) +
+                            std::pow(x2 - yrough_ptr[ii], 2));
+                        if (radius < residual) {
+                            residual = radius;
+                            roughz0 = z0rough_ptr[ii];
+                        }
+                        if (radius < dx[0]) {
+                            break;
+                        }
+                    }
+                    levelz0(i, j, k, 0) = roughz0;
+                });
+            amrex::ParallelFor(
+                vbx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                    // Terrain Height
+                    amrex::Real residual = 10000;
+                    amrex::Real terrainHt = 0.0;
+                    const amrex::Real x1 = prob_lo[0] + (i + 0.5) * dx[0];
+                    const amrex::Real x2 = prob_lo[1] + (j + 0.5) * dx[1];
+                    const amrex::Real x3 = prob_lo[2] + (k + 0.5) * dx[2];
+                    for (unsigned ii = 0; ii < terrainSize; ++ii) {
+                        const amrex::Real radius = std::sqrt(
+                            std::pow(x1 - xterrain_ptr[ii], 2) +
+                            std::pow(x2 - yterrain_ptr[ii], 2));
+                        if (radius < residual) {
+                            residual = radius;
+                            terrainHt = zterrain_ptr[ii];
+                        }
+                    }
+                    levelDrag(i, j, k, 0) = 0.0;
+                    if (x3 > terrainHt && k > 0 &&
+                        levelBlanking(i, j, k - 1, 0) == 1) {
+                        levelDrag(i, j, k, 0) = 1.0;
```

Comment:
are terrainBlank and terrainDrag just intended to be 1 or 0? If so, they're better suited to be int fields (fields composed of integers) instead of ordinary fields.

Reply:
I need to access this field inside the turbulence class to blank out turbulent viscosity values. This was one of the reasons I kept it as a field since I can initialize to a random field within the turbulence class.

Reply:
understood. there is a better way to do it, and it will allow you to use an int field instead of a field. the treatment of reference_density (and rho0_arr) in equation_systems/icsn/source_terms/Gravity_Forcing.cpp shows how to do it. In the mf loop, if the field exists, you use it; otherwise, you can reference amrex::Array4<int>().

Reply:
When I declared the blanking field as an int field, it compiled fine. However, when running, I got an error message saying the variable is not available. I am not sure if int field has some special access methods. Currently, changed it back to a regular field.

Reply:
This doesn't surprise me. You need to do more than just change the declaration; you have to change the use of the field as well -- which just means you have to replace the field-related calls with int-field-related ones. In your case, you need to replace get_field with get_int_field. That kind of error isn't something the compiler can pick up. The field_exists call will need to be replaced with int_field_exists for terrainBlank. You'll also need to replace register_io_var with register_output_int_var.